### PR TITLE
Issue 15653 - IFTI fails for immutable parameter

### PR DESCRIPTION
**Pull request by 9rnsr**

### CODE REVIEW

File: src/dtemplate.d [hunk 1]
```diff
@@ -4059,27 +4062,46 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
             {
                 result = m;
             }
+            else if (e.type.toBasetype().isintegral() ||
```
Comment (yebblies) [inline src/dtemplate.d]: What about static arrays?  Can't this be checked using type converts to tailconst(type)?

Comment (9rnsr) [inline src/dtemplate.d]: Yes, also for `Tsarray` and `Tstruct`, we have to check top-qualifier conversion. Will do it later.

Comment (9rnsr) [inline src/dtemplate.d]: OK, done.

In D, only three types (`Tclass`, `Taarray`, and `Tstruct` with `hasPointers() == true`) cannot support the top-qualifier conversion. If `e.type` is not one of them, we IFTI can deduce corresponding type parameter.

**Comment by JinShil** [comment]
Revived at #7977

**Pull request closed without merging**